File: 300-399/354-russian-doll-envelopes.py
from typing import List
from bisect import bisect_left

class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        envelopes.sort()
        
        result = 0
        dp = [0] * len(envelopes)
        # O(n^2) LIS
        for i, largerEnv in enumerate(envelopes):
            dp[i] = 1
            for j, env in enumerate(envelopes[:i]):
                if largerEnv[0] > env[0] and largerEnv[1] > env[1]:
                    dp[i] = max(dp[i], dp[j] + 1)
            result = max(result, dp[i])
        return result      

    # https://leetcode.com/problems/russian-doll-envelopes/discuss/1134011/JS-Python-Java-C%2B%2B-or-Easy-LIS-Solution-w-Explanation  
    def maxEnvelopes(self, E: List[List[int]]) -> int:
        E.sort(key=lambda x: (x[0], -x[1]))
        dp = []
        for _,height in E:
            left = bisect_left(dp, height)
            if left == len(dp): dp.append(height)
            else: dp[left] = height
        return len(dp)
        

File: 300-399/354-russian-doll-envelopes.go
package main

import "sort"

// Slow: O(n^2) time complixity
func maxEnvelopes(envelopes [][]int) int {
	sort.Slice(envelopes, func(i, j int) bool {
		ei, ej := envelopes[i], envelopes[j]
		if ei[0] == ej[0] {
			return ei[1] < ej[1]
		}
		return ei[0] < ej[0]
	})

	var (
		dp     = make([]int, len(envelopes))
		result int
	)
	for i, env := range envelopes {
		dp[i] = 1
		for j, smallEnv := range envelopes[:i] {
			if env[0] > smallEnv[0] && env[1] > smallEnv[1] && dp[i] < dp[j]+1 {
				dp[i] = dp[j] + 1
			}
		}
		if result < dp[i] {
			result = dp[i]
		}
	}
	return result
}
